hazelcast/src/main/java/com/hazelcast/jet/impl/processor/AsyncTransformUsingServiceUnorderedP.java

```diff
@@ -83,18 +83,44 @@
     In the resultQueue we also store the last received WM value for each key.
 
     Separately, we track watermark counts for each key, which we increment for each WM key when an event is received,
-    and decrement when a response is processed. The count is the count of events received _before_ that WM, since
-    the previous WM. When the count gets to 0, we know we can emit the watermark, because all the responses
+    and decrement when a response is processed. The count is the count of events received _since_ that WM, _before_
+    the next WM. When the count gets to 0, we know we can emit the next watermark, because all the responses
     for events received before it were already sent.
+
+    Snapshot contains in-flight elements at the time of taking the snapshot.
+    They are replayed when state is restored from the snapshot, so we get only at-least-once guarantee.
```

Comment:
The guarantee is still exactly-once, but under the hood the processing is retried at least once, but that doesn't break the guarantee.

Reply:
updated comment